### DISCUSSION TITLE
*Why* is there no `ConvertFrom-XML` commandlet?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [PowerShell/PowerShell#24286] [Q&A] [answered]
**Question**

As originally asked by another at [SO](https://stackoverflow.com/questions/3242995/convert-xml-to-psobject#comment106240484_44852653):

1. > [How to convert XML to PSObject?](https://stackoverflow.com/review/suggested-edits/36434052)
1. > You can do this pretty easily with XPath.
1. > Why isn't this a normal built-in function?

I ask because [`github.com/MicrosoftDocs/PowerShell-Docs/blob/d135dcb0d790368ce0ff5a4a07c2aacc24df5419/reference/7.4/Microsoft.PowerShell.Utility/ConvertTo-Xml.md#convertto-xml`](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/d135dcb0d790368ce0ff5a4a07c2aacc24df5419/reference/7.4/Microsoft.PowerShell.Utility/ConvertTo-Xml.md#convertto-xml:~:text=ConvertTo%2DXml-,ConvertTo%2DXml,-SYNOPSIS) exists, yet no `ConvertFrom` counterpart exists, which is unusual, especially since [`.*-JSON`](https://regex101.com/r/T0BBDq/1) has both.

**(Lack of) Duplicates**

[`github.com/PowerShell/PowerShell/discussions?discussions_q=is%3Aopen+ConvertFrom-XML`](https://github.com/PowerShell/PowerShell/discussions?discussions_q=is%3Aopen+ConvertFrom-XML#:~:text=is:open-,ConvertFrom%2DXML) doesn't appear to return anything relevant.

**Comment by rhubarb-geek-nz** [comment, marked as answer]
Given this is a couple of months old and no answers, I'll bite. This is purely opinion but I expect that it is because (a) it isn't really needed (b) it can't be done well.

I expect that the cast is good enough for most cases.

```
[xml]$NuGetConfig = Get-Content -LiteralPath $NuGetConfigFile
```

Now XML is a nightmare, if you have namespaces, attributes, array elements with only one item, any type other than text then a a simple conversion will give you a very poor representation.

The existing mechanism gives you an XML DOM Document, where the XML DOM API is already a cross language/platform API and much more generically useful that converting to a PSCustomObject. PowerShell lets you dive directly into named elements if there are no name clashes, and you can use the DOM api directly to navigate.

If you want ease of use then use JSON, if you are using XML because you have to integrate with another system then the DOM API will give you the control you need.

Searching for "ConvertFrom-XML" in PSGallery gives nine results which shows there is no one right answer for how to do this.

Now given PowerShell is based on dotnet, perhaps the answer is [XML Serialization with XML Web Services](https://learn.microsoft.com/en-us/dotnet/standard/serialization/xml-serialization-with-xml-web-services) but even that needs a WSDL/Schema and a [tool](https://learn.microsoft.com/en-us/dotnet/core/additional-tools/dotnet-svcutil-guide?tabs=dotnetsvcutil2x) to generate the object representation.

So my simple anwser is that it cannot be both simple and done well enough for all cases. The PSGallery has enough implementations for you to find one that meets your needs, if not write one and add it to the PSGallery or simply use the XML cast and it will also work on PowerShell 5.

Even the existing ConvertTo-JSON and ConvertFrom-JSON have problems round-tripping data with no changes in structure. I don't see a a ConvertFrom-XML succeeding where ConvertFrom-JSON fails.

**Reply by roke-julian-lockhart** [reply]
Thank you again, @rhubarb-geek-nz.
